### PR TITLE
Feat curl conversion

### LINKED ISSUE #43: Log Failed Request cURL
Curl's are very helpful to investigate failing requests in postman

### PR DESCRIPTION
**Pull request by NoahKamara**
Implement an interceptor that outputs requests as a cURL command when a condition is met

This also implements a method `Request.curl(sortHeaders:) -> String` which turns a request into a cURL command to make it easy to implement custom interceptors that output curl

- Fixes #43

### CODE REVIEW

File: PapyrusCore/Sources/Request.swift [hunk 1]
```diff
@@ -6,3 +6,44 @@ public protocol Request {
...
+        // Add headers
+        if sortedHeaders {
+            for (key, value) in headers.sorted(by: { $0.key < $1.key }) {
+                components.append("-H '\(key): \(value)'")
+            }
+        } else {
+            for (key, value) in headers {
+                components.append("-H '\(key): \(value)'")
+            }
+        }
```
Comment (joshuawright11) [inline PapyrusCore/Sources/Request.swift:39]: Slightly cleaner option that I believe works:

```suggestion
        let headerOptions = headers.map { "-H '\($0): \($1)'" }
        components += sortedHeaders ? headerOptions.sorted() : headerOptions
```

File: PapyrusCore/Sources/Request.swift [hunk 2]
```diff
@@ -6,3 +6,44 @@ public protocol Request {
...
+    /// Create a cURL command from this instance
+    ///
+    /// - Parameter sortedHeaders: sort headers in output
+    /// - Returns: cURL Command
+    func curl(sortedHeaders: Bool = false) -> String {
+        let lineSeparator = " \\\n"
+        var components = [String]()
+
+        // Add URL on same line
+        if let url = self.url {
```
Comment (joshuawright11) [inline PapyrusCore/Sources/Request.swift:21]:
```suggestion
        if let url {
```

File: PapyrusCore/Sources/Request.swift [hunk 3]
```diff
@@ -6,3 +6,44 @@ public protocol Request {
...
+            }
+        } else {
+            for (key, value) in headers {
+                components.append("-H '\(key): \(value)'")
+            }
+        }
+
+        // Add body
+        if let bodyData = body {
+            let bodyString = String(data: bodyData, encoding: .utf8) ?? ""
```
Comment (joshuawright11) [inline PapyrusCore/Sources/Request.swift:43]:
```suggestion
        if let body {
            let bodyString = String(data: body, encoding: .utf8) ?? ""
```

File: PapyrusCore/Tests/CurlTests.swift [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    func testInterceptorOnErrorNoError() async throws {
+        var req = RequestBuilder(baseURL: "foo/", method: "GET", path: "/baz")
+        req.addQuery("Hello", value: "There")
+        req.addHeader("High", value: "Ground")
+
+        let request = try TestRequest(from: req)
+
+        let normalizedCurl = request.curl(sortedHeaders: true)
```
Comment (joshuawright11) [inline PapyrusCore/Tests/CurlTests.swift:176]: unused - delete?

File: PapyrusCore/Sources/Interceptors/CurlInterceptor.swift [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+/// An `Interceptor` that logs requests based on a condition
+public struct CurlLogger {
+    public enum Condition {
+        case always
+
+        /// only log when the request encountered an error
+        case onError
+    }
+
+    let loggingFn: (String) -> Void
```
Comment (joshuawright11) [inline PapyrusCore/Sources/Interceptors/CurlInterceptor.swift:12]: nit: would you mind making this `log` or `logHandler`? Slightly more in line with other closure names in codebase.

**Review by joshuawright11 — APPROVED** [review]
Wow, this is awesome! 

Great work @NoahKamara 👏👏

Couple minor nits then feel free to merge, I'll cut a release when it's in.

File: PapyrusCore/Sources/Request.swift (hunk 1 above)
Comment (NoahKamara) [reply PapyrusCore/Sources/Request.swift:39]: very nice. this has the added benefit of also sorting by value two headers have the same key

File: PapyrusCore/Tests/CurlTests.swift (hunk 4 above)
Comment (NoahKamara) [reply PapyrusCore/Tests/CurlTests.swift]: ah yes. left by mistake after copy-paste

File: PapyrusCore/Sources/Interceptors/CurlInterceptor.swift (hunk 5 above)
Comment (NoahKamara) [reply PapyrusCore/Sources/Interceptors/CurlInterceptor.swift]: fixed

**Comment by NoahKamara** [comment]
> Couple minor nits then feel free to merge, I'll cut a release when it's in.

I think i resolved all the comments but im not a collaborator so you're gonna have to do the merging :)

**Comment by joshuawright11** [comment]
Amazing work @NoahKamara!

**Pull request merged**